### PR TITLE
Transform improvements and mixed sparse-dense transform

### PR DESCRIPTION
**Pull request by SimonHeybrock**
This is the main step towards #676. Changes are:

- Refactor `visit.h` to support arbitrary argument counts.
- Refactor `transform.h` to support arbitrary argument counts in more places. This is not complete yet, but most parts are there now.
- Support `span` as element type in `Variable`. This is then used to demonstrate have we can use `transform` to implement an operation between sparse data and a histogram (see #676 for details). This is a poor-man's version of the transform with core dimensions. Unless performance turns out to be an issue, it is probably sufficient.
- Various other cleanup and restructuring.

### CODE REVIEW

File: core/include/scipp/core/transform_common.h [hunk 1]
```diff
@@ -56,6 +60,18 @@ static constexpr auto dimensionless_unit_check_return =
...
 
+namespace transform_flags {
+/// Add this to overloaded operator to indicate that the operation does not
+/// produce output with variances, even if the inputs contain variances.
+static constexpr auto no_variance_output = []() {};
+using no_variance_output_t = decltype(no_variance_output);
+
+template <int N> static constexpr auto expect_no_variance_arg = []() {};
+template <int N>
+using expect_no_variance_arg_t = decltype(expect_no_variance_arg<N>);
```
Comment (igudich) [inline core/include/scipp/core/transform_common.h]: Would this work: 
```cpp
static constexpr auto dummy_lambda = []() {};
using dummy_lambda_t = decltype(dummy_lambda);
using no_variance_output_t = dummy_lambda_t;
template <int N>
using expect_no_variance_arg_t = dummy_lambda_t;
```

Comment (SimonHeybrock) [reply core/include/scipp/core/transform_common.h]: No it would not work, since `transform` checks for the flag based on its type, so `N` must be part of the type. Having it part of the typedef is not sufficient.

File: core/test/transform_sparse_and_dense_test.cpp [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+  auto dense_view = makeVariable<span<double>>({Dim::Y, 2}, spans);
+
+  const auto result = transform<
+      pair_custom_t<std::pair<sparse_container<double>, span<double>>>>(
+      var, dense_view,
+      overloaded{[](const auto &a, const auto &) { return a; },
+                 [](const units::Unit &a, const units::Unit &) { return a; },
+                 transform_flags::expect_no_variance_arg<0>,
+                 transform_flags::expect_no_variance_arg<1>});
+}
```
Comment (igudich) [inline core/test/transform_sparse_and_dense_test.cpp]: `EXPECT` something in here?

File: core/test/transform_sparse_and_dense_test.cpp [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+            return c;
+          }});
+  EXPECT_TRUE(out.hasVariances());
+  const auto out_vals = out.sparseValues<float>();
+  EXPECT_TRUE(equals(out_vals[0], {2, 3, 3}));
+  EXPECT_TRUE(equals(out_vals[1], {2, 2, 3, 1}));
+  const auto out_vars = out.sparseVariances<float>();
+  EXPECT_TRUE(equals(out_vars[0], {0.3f, 0.4f, 0.4f}));
+  EXPECT_TRUE(equals(out_vars[1], {0.3f, 0.3f, 0.4f, 1.0f}));
+}
```
Comment (igudich) [inline core/test/transform_sparse_and_dense_test.cpp]: Do we need `sparse_time_dense` as a function in API? In other case I would say this test is  to sophisticated and complex.

Also no tests for incorrect arguments and corner cases.

File: core/variable_instantiate_view_elements.cpp [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+#include "scipp/core/variable.h"
+#include "scipp/core/variable.tcc"
+
+namespace scipp::core {
+
+// Variable's elements are views into other variables. Used internally for
+// implementing functionality for sparse data combined with dense data using
+// transform.
+INSTANTIATE_VARIABLE(span<double>)
+INSTANTIATE_VARIABLE(span<float>)
```
[the remark this answers is not in the document]
Comment (SimonHeybrock) [reply core/variable_instantiate_view_elements.cpp]: Yes I think so, unless we want to include `variable.tcc` everywhere where we use it. Also, it should avoid multiple instantiations with the same element type, when used in multiple places.

File: core/include/scipp/core/transform.h [hunk 5]
```diff
@@ -425,13 +348,15 @@ template <class Op> struct Transform {
                   dims, default_init_elements)
             : makeVariable<element_type_t<Out>>(dims, default_init_elements);
     auto &outT = static_cast<VariableConceptT<Out> &>(out.data());
-    do_transform(as_view{*handles, dims}..., outT, op);
+    do_transform(op, outT, std::tuple<>(), as_view{*handles, dims}...);
```
[the remark this answers is not in the document]
Comment (SimonHeybrock) [reply core/include/scipp/core/transform.h]: What is the confusing part?

File: core/test/transform_sparse_and_dense_test.cpp (hunk 2 above)
Comment (SimonHeybrock) [reply core/test/transform_sparse_and_dense_test.cpp]: See below.

File: core/test/transform_sparse_and_dense_test.cpp (hunk 3 above)
Comment (SimonHeybrock) [reply core/test/transform_sparse_and_dense_test.cpp]: Sorry, I should have clarified this more in the PR description: Essentially this is a prototype for what I will implement for #676 in the next step. The code in these tests will be split into two parts:
1. Helper to generate the `Variable` with `T=span`.
2. Operator that does the multiplication (and division).

This test will likely just disappear, or be replaced by something more meaningful. For now it was mainly to check that we can compile what needs to be compiled.

File: core/include/scipp/core/transform.h [hunk 6]
```diff
@@ -350,56 +277,52 @@ struct TransformSparse {
...
+          op, std::forward<Out>(out),
+          std::tuple_cat(processed, std::tuple(ValuesAndVariances{vals, vars})),
+          args...);
     }
   } else {
-    transform_elements(op, out_val, a_val, b_val);
+    do_transform(op, std::forward<Out>(out),
+                 std::tuple_cat(processed, std::tuple(vals)), args...);
   }
 }
```
[the remark this answers is not in the document]
Comment (SimonHeybrock) [reply core/include/scipp/core/transform.h]: I had tried to find a way doing it using fold expression, but found it too complicated to handle the failure cases. I will have another look based on your suggestion.

File: core/include/scipp/core/transform.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (igudich) [reply core/include/scipp/core/transform.h]: Does `view` tends to be immutable?

[the remark this answers is not in the document]
Comment (SimonHeybrock) [reply core/include/scipp/core/transform.h]: No, look e.g., at the views in C++20.
